pkg/metrics/metrics.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package metrics contains global structures related to metrics collection
+// cert-manager exposes the following metrics:
+// certificate_requests{name, namespace, issuer_name, issuer_scope, issuer_type, operation, result}
+// certificate_expiry_time_seconds{name, namespace}
+package metrics
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/gorilla/mux"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	// Namespace is the namespace for cert-manager metric names
+	Namespace                              = "certmanager"
+	prometheusMetricsServerAddress         = "0.0.0.0:9402"
+	prometheusMetricsServerShutdownTimeout = 10 * time.Second
+	prometheusMetricsServerReadTimeout     = 8 * time.Second
+	prometheusMetricsServerWriteTimeout    = 8 * time.Second
+	prometheusMetricsServerMaxHeaderBytes  = 1 << 20
+)
+
+type Metrics struct {
+	http.Server
+
+	registry                     *prometheus.Registry
+	CertificateRequests          *prometheus.CounterVec
+	CertificateExpiryTimeSeconds *prometheus.GaugeVec
+}
+
+func New() *Metrics {
+
+	router := mux.NewRouter()
+
+	// Create server and register prometheus metrics handler
+	s := &Metrics{
+		Server: http.Server{
+			Addr:           prometheusMetricsServerAddress,
+			ReadTimeout:    prometheusMetricsServerReadTimeout,
+			WriteTimeout:   prometheusMetricsServerWriteTimeout,
+			MaxHeaderBytes: prometheusMetricsServerMaxHeaderBytes,
+			Handler:        router,
+		},
+		registry: prometheus.NewRegistry(),
+		CertificateRequests: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: Namespace,
+				Name:      "certificate_requests",
+				Help:      "Number of certificate requests",
+			},
+			[]string{"name", "namespace", "issuer_name", "issuer_kind", "issuer_type", "operation", "result"},
+		),
+		CertificateExpiryTimeSeconds: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: Namespace,
+				Name:      "certificate_expiry_time_seconds",
+				Help:      "Number of seconds after January 1, 1970 UTC that the certificate will expire",
```

Comment:
You don't want to depend on the cert-manager container's time.Now(). Rather having a static value of the expiry time and do the remaining time calculation on the prometheus server. Unix time representation is used in the node exporter like that: https://github.com/prometheus/node_exporter/blob/a98067a294bd6cebde76d3078ba85b73b7166a74/collector/stat_linux.go#L57

Docs don't define a best practice but I think something in the node_exporter should be considered as such

Reply:
My feel is the metrics endpoints are for prometheus and not directly for humans, so the metrics exposed should be whatever form/unit is most useful for creating useful expressions within prometheus. 

Prometheus provides [`time()`](https://prometheus.io/docs/prometheus/latest/querying/functions/#time()) which follows the unix convention of seconds since 1970. So if you want your preferred graph @munnerz is it just `(metric - time())` in the prometheus query, whether in prometheus UI or in downstream Graphana dashboards.

  "time() returns the number of seconds since January 1, 1970 UTC."

That said, these is no limit to exposing multiple metrics for the same thing, so you could have both the absolute and relative time for expiry (according to `cert-manager`) as metrics.